src/test/python/deep_qa_test/layers/attention/attention_test.py

```diff
@@ -1,13 +1,92 @@
 # pylint: disable=no-self-use,invalid-name
 
-import numpy
+import numpy as np
 from numpy.testing import assert_almost_equal
 from keras.layers import Embedding, Input
 from keras.models import Model
-
+import keras.backend as K
 from deep_qa.layers.attention.attention import Attention
 
 class TestAttentionLayer:
+    # pylint: disable=no-self-use
+    def test_no_mask(self):
+        vector_length = 3
+        matrix_num_rows = 2
+
+        vector_input = Input(shape=(vector_length,),
+                             dtype='float32',
+                             name="vector_input")
+        matrix_input = Input(shape=(matrix_num_rows, vector_length),
+                             dtype='float32',
+                             name="matrix_input")
+        similarity_softmax = Attention()([vector_input, matrix_input])
+        model = Model([vector_input, matrix_input],
+                      similarity_softmax)
+
+        # Testing general non-batched case.
+        vector = np.array([[0.3, 0.1, 0.5]])
+        matrix = np.array([[[0.6, 0.8, 0.1], [0.15, 0.5, 0.2]]])
+
+        result = model.predict([vector, matrix])
+        assert_almost_equal(result, np.array([[0.52871835, 0.47128162]]))
+
+        # Testing non-batched case where inputs are all 0s.
+        vector = np.array([[0, 0, 0]])
+        matrix = np.array([[[0, 0, 0], [0, 0, 0]]])
+
+        result = model.predict([vector, matrix])
+        assert_almost_equal(result, np.array([[0.5, 0.5]]))
+
+    def test_masked(self):
+        # Testing general masked non-batched case.
+        vector = K.variable(np.array([[0.3, 0.1, 0.5]]))
+        matrix = K.variable(np.array([[[0.6, 0.8, 0.1], [0.15, 0.5, 0.2], [0.1, 0.4, 0.3]]]))
+        mask = K.variable(np.array([[1.0, 0.0, 1.0]]))
+        result = K.eval(Attention().call([vector, matrix], mask=["_", mask]))
```

Comment:
a masking layer that lets us mask inputs that aren't just single values would be very useful, we should consider writing one...